R/call_standarise_formals.R

```diff
@@ -15,25 +15,135 @@ call_standardise_formals <- function(code, env = rlang::current_env(), include_d
   # out default formals. For primitives like mean, we're unable to distinguish
   # between mean() and mean.default()
   if (is_false(include_defaults) || is_infix(code) || is.primitive(fn)) {
-    return(rlang::call_standardise(code, env = env))
+    return(call_standardise_keep_partials(code, env = env))
   }
 
   fmls <- rlang::fn_fmls(fn)
   args_default <- fmls[!vapply(fmls, is.symbol, logical(1), USE.NAMES = FALSE)]
 
-  code_std <- rlang::call_standardise(code, env = env) # order and label existing params
+  # order and label existing params
+  code_std <- call_standardise_keep_partials(code, env = env)
 
-  # get arguments passed from user
+  # get named arguments passed from user
   args_user <- rlang::call_args(code_std)
+  args_user <- args_user[nzchar(names(args_user))]
 
-  args_default_missing <- setdiff(names(args_default), names(args_user))
+  args_default_missing <- names(args_default)[
+    !grepl(paste0("^", names(args_user), collapse = "|"), names(args_default))
+  ]
   if (length(args_default_missing) == 0) {
     return(code_std)
   }
 
   ## Add implicit default args to the call
-  rlang::call_standardise(
+  call_standardise_keep_partials(
     rlang::call_modify(code_std, !!!args_default[args_default_missing]),
     env = env
   )
 }
+
+call_standardise_keep_partials <- function(code, env = rlang::caller_env()) {
+  # If the function from the user code cannot be found, early exit because
+  #   standardization is impossible
+  fn <- tryCatch(call_fn(code, env), error = as.null)
+  if (is.null(fn)) return(code)
+
+  tryCatch(
+    rlang::call_match(code, fn),
+    error = function(e) {
+      # Check that error is caused by an ambiguous partial argument
+      # If not, return the code unaltered
+      pattern <- paste0(
+        "\\Q",
+        gsub(
+          pattern = "%d",
+          replacement = "\\E\\d+\\Q",
+          x = gettext("argument %d matches multiple formal arguments", domain = "R"),
+          fixed = TRUE
+        ),
+        "\\E"
+      )
+      if (!grepl(pattern, e$message)) {
+        return(code)
+      }
```

Comment:
Will this end up swallowing other kinds of errors?

Reply:
Yes, the idea is that if we have any kind of parsing error besides a "argument x matched multiple formal arguments" error, it will just treat the code as unstandardizable and return it as is. Are there any other errors that we'd like to try to recover from?

Reply:
That makes sense, I just wanted to make sure we've accounted for upstream places that might want the standardized code but I think they're caught by other mistake detectors... here's how I convinced myself.

```r
dummy <- function(aaa = 1, aaaa = 2) {
  aa + aaa * 10
}

grade_this_code()(mock_this_exercise("dummy(a = 1, aa = 2)", "dummy(2, 1)"))
#> <gradethis_graded_this_code: [Incorrect]
#>   `dummy()` accepts an argument named `"aaa"`. More than one of your
#>   argument names in `dummy()` will be matched to `"aaa"`, which will
#>   cause an error. Try writing out the full argument names. Try it
#>   again. Perseverence is the key to success.
#> >

call_standardise_keep_partials(quote(dummy(a = 1, aa = 2)))
#> dummy(a = 1, aa = 2)
```